Refactor to accept concurrency and make it more testable

Tavernari opened the pull request:
Hi @dhargrow,
 
I was thinking about how to explain through comments the changes that I believe could improve the implementation, so instead of writing a ton of text, I decided to propose a refactor based on your implementation.
 
The key changes in this proposal for refactoring are:
 
- Making it even more testable by separating concerns, such as storage from session management.
- Preparing the SessionManager and Storage for Swift Concurrency features.
- Updating Sessions to be more "intelligent" and enhancing usability and domain rules.
 
 
Feel free to comment if you have any questions or if you would like to make changes; this is just a proposal for improving the implementation.

Tavernari committed:
Refactor to accept concurrency and make it more testable

dhargrow commented on Package.swift:
Should the readme reflect the platforms this is available for too?

Tavernari replied:
Yes, it is true!
Done on this commit [Add platform requirements section to README.md](https://github.com/Tavernari/SwiftCuckoo/pull/20/commits/9aa92409bcb6be126ac8c6b5f288a708cc2a138f)

dhargrow commented on Sources/SwiftCuckoo/Session.swift:
```diff
@@ -2,37 +2,74 @@ import Foundation
+    /// - Parameter date: The date to stop the session. Defaults to the current date.
+    /// - Throws: An error of type `Error` if:
+    ///   - The session has not been started (must start before stopping).
+    public mutating func stop(on date: Date = Date()) throws {
```
I was thinking of going this route! I haven't played around with mutating functions much before this. However the reason we need the function to be mutating is because structs aren't mutable like classes due to being value types?

Tavernari replied:
Structs are value types instead of references. So you can change the original object, only copy (There is an internal mechanism to avoid unnecessary copies a.k.a copy-on-write). 
In that case, mutating methods will say to the compiler that the method will change internal state and the compiler will complain if it is a constant because it will need to make a copy and replace the current value.

dhargrow reviewed (approved):
This is good, some of the concurrency stuff is a little new to me like AnyActor (which is deprecated)  & Sendable

Tavernari committed:
Change protocol conformance from AnyActor to Sendable

Replaced `AnyActor` with `Sendable` in the `SessionManagerStorage` protocol to align with concurrency or safety requirements.

Tavernari committed:
Fix typo in README.md

Corrected a typo in the "Change Notifications" section by replacing "ppply" with "apply."

Tavernari committed:
Add platform requirements section to README.md

Included a new "Platforms" section in the README.md specifying minimum platform versions for macOS (10.15+) and iOS (13.0+).

Tavernari commented:
> This is good, some of the concurrency stuff is a little new to me like AnyActor (which is deprecated) & Sendable

Thanks, I replaced AnyActor by Sendable on this commit [Change protocol conformance from AnyActor to Sendable](https://github.com/Tavernari/SwiftCuckoo/pull/20/commits/c00a64852536366de637f7830cf1885308f4a2a9)

Concurrency is tricky and also important, so since you are starting, I suggest you learn Swift Concurrency but also study GCD through this [documentation]( https://developer.apple.com/documentation/DISPATCH)

For now we are going to use a modern approach with [Swift Concurrency](https://docs.swift.org/swift-book/documentation/the-swift-programming-language/concurrency/)
